pkg/controller/machine/controller.go

```diff
@@ -194,9 +194,14 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
 }
 
 func (r *ReconcileMachine) getCluster(ctx context.Context, machine *clusterv1.Machine) (*clusterv1.Cluster, error) {
+	if machine.Labels["cluster"] == "" {
```

Comment:
The reason this is using a label is still so that people can fake a cluster-id if they don't want to use the cluster object?

Reply:
Someone brought up that labels were preferred during our weekly meeting. This wouldn't make the cluster optional, but users might use a dummy cluster if they don't want to make sure of the Cluster type.

Reply:
The reason why labels and/or annotations where mentioned as being preferred is because they are a much weaker link between the objects.

Reply:
is "cluster" label become a required field in machineSpec? i am debating whether we should provide flexibility here, if "cluster" label not been set, use namespace to find cluster.

Reply:
I think we should either be explicit or return no cluster. Returning the first cluster in a namespace (which is the current behavior) might bring unexpected behavior for consumers.

Reply:
If we were to use labels as the governing mechanism to select a cluster, I would imagine using something like a LabelSelector in the machine spec. e.g.
```
/// [MachineSetSpec]
// MachineSetSpec defines the desired state of MachineSet
type MachineSetSpec struct {
  ...
  ClusterSelector metav1.LabelSelector `json:"clusterSelector"`
}
```

Reply:
Regarding selectors: I think that's a mechanism to select a set of things and not necessarily a single thing. It would make more sense having a selector in the Cluster type to select machines. But this doesn't help much when going from machine to cluster.

If you're using a single label, you could consider a non-controller owner reference in the metadata. If you wanted to take it even further, machine deployments and machine sets could have this too. This has the added benefit of `kubectl delete -n <namespace> cluster <clustername>` doing a cascading deletion across the resources.

Reply:
@maisem I'm not sure how a cluster selector would work in this case. The label selector wouldn't be able to select a cluster by name given that the name is a field.

As @krousey pointed out, label selector are more suited when querying list of resources. The problem that I see with `ownerRef` though is that's a larger change and impacts delete which I think it should be revisited during the next iteration.

For this cycle I think having a prefixed label in the machine that points to a cluster name should fix the immediate issue and won't bring in many changes to the current logic and providers. 

I'll open issues to address these long-term concerns, how does that sound?

Reply:
> This has the added benefit of `kubectl delete -n <namespace> cluster <clustername>` doing a cascading deletion across the resources.

@krousey In the case of the AWS provider, the cloud-provider objects managed by the Cluster object are unable to be deleted while there are existing Machines, so the cascading deletion would not happen.

Reply:
@detiber Sounds like you want foreground cascading deletion then. https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion

@vincepri This has been an issue for a while. The `ownerRef` change is something that would need a wider discussion too.